primer/src/Primer/Action.hs

```diff
@@ -230,6 +244,49 @@ applyActionsToTypeSig smartHoles imports (mod, mods) (defName, def) actions =
             -- In this case we just refocus on the top of the type.
             z -> maybe unwrapError pure (focusType (unfocusLoc z))
 
+applyActionsToField ::
+  (MonadFresh ID m, MonadFresh NameCounter m) =>
+  SmartHoles ->
+  [Module] ->
+  (Module, [Module]) ->
+  (Name, ValConName, Int, ASTTypeDef TypeMeta) ->
+  [Action] ->
+  m (Either ActionError ([Module], TypeZ))
+applyActionsToField smartHoles imports (mod, mods) (tyName, conName', index, tyDef) actions =
+  runReaderT
+    go
+    (buildTypingContextFromModules (mod : mods <> imports) smartHoles)
+    & runExceptT
+  where
+    go :: ActionM m => m ([Module], TypeZ)
+    go = do
+      (valCons, zt) <-
+        (maybe (throwError $ InternalFailure "applyActionsToField: con name not found") pure =<<) $
+          flip (findAndAdjustA' ((== conName') . valConName)) (astTypeDefConstructors tyDef) \(ValCon _ ts) -> do
+            (t, zt) <-
+              maybe (throwError $ InternalFailure "applyActionsToField: con field index out of bounds") pure
+                =<< flip (adjustAtA' index) ts \fieldType -> do
```

Comment:
I don't really like `adjustAtA'` and `findAndAdjustA'`. I feel like there _should_ be some way to use an appropriate `Applicative`, like `Compose (,)`, rather than having to create a whole new helper to thread the extra data along. But I couldn't work out how to make this work. Any ideas?

Reply:
As things stand, there are two difficulties.

Firstly, from the type of `adjustAtA`, we have no reason to believe that the function will be called any particular number of times, thus whatever `Applicative` we use needs to handle the possibility of multiple `z`s. There are three ways to improve this: either weaken the `Applicative` constraint to something like `Pointed` (but this steps out of the standard hierarchy, so may be awkward to use); define something like `adjustAtA0 :: Functor f => Int -> (a -> f a) -> [a] -> Maybe (f [a])` (essentially the same idea, but makes it explicit in the return type; this new function can be used to define `adjustAtA`); or keep the current type and use `First`.

The second difficulty is the chosen order of the tuple. We have no way to say `(,z)` at type level, only `(z,)`, written as `(,) z`. If you change the order of the tuple it is then possible to define `adjustAtA'` in terms of `adjustAtA`.

This boils down to one of the following:
```
adjustAtA0 :: Functor f => Int -> (a -> f a) -> [a] -> Maybe (f [a])
adjustAtA0 n f xs = case splitAt n xs of
  (a, b : bs) -> Just $ f b <&> \b' -> a ++ [b'] ++ bs
  _ -> Nothing

-- | Like 'adjustAt', but in an 'Applicative'.
adjustAtA :: Applicative f => Int -> (a -> f a) -> [a] -> f (Maybe [a])
adjustAtA n f xs = sequenceA $ adjustAtA0 n f xs

adjustAtA' :: forall f a z . Applicative f => Int -> (a -> f (z, a)) -> [a] -> f (Maybe (z, [a]))
adjustAtA' n f = traverse getCompose . adjustAtA0 n f'
  where f' :: a -> Compose f ((,) z) a
        f' = Compose . f
```

```
adjustAtA' :: forall f a z . Applicative f => Int -> (a -> f (z, a)) -> [a] -> f (Maybe (z, [a]))
adjustAtA' n f = fmap (_ . first getFirst) . getCompose . adjustAtA n f'
  where f' :: a -> Compose f ((,) (First z)) a
        f' = Compose . fmap (first (First . Just)) . f
        g :: Applicative g => (g a, g b) -> g (a, b)
        g (x,y) = liftA2 (,) x y
```